[Docs] Data model (#1374)
wasp-lang/wasp: web/docs/database/operations/queries.md

---
title: Queries
---

import { Required } from '@site/src/components/Required';

We'll explain what Queries are and how to use them. If you're looking for a detailed API specification, skip ahead to the [Options Reference](#options-reference).

You can use Queries to fetch data from the server. They shouldn't modify the server's state.
Fetching all comments on a blog post, a list of users that liked a video, information about a single product based on its ID... All of these are perfect use cases for a Query.

> :::tip

Reviewer: I would maybe remove this note 🤷  let them read, they skim over the similar bits?

Author: Talked about it, decided to leave it for now.

Queries are fairly similar to Actions in terms of their API.
Therefore, if you're already familiar with Actions, you might find reading the entire guide repetitive.

We instead recommend skipping ahead and only reading [the differences between Queries and Actions](/docs/database/operations/actions#differences-between-queries-and-actions), and consulting the [Options Reference](#options-reference) as needed.
:::

## Working with Queries

> Queries are declared in Wasp and implemented in NodeJS. Wasp executes Queries within the server's context but also generates the code that lets you call them from anywhere in your code (client or server) via the same interface.
> In other words, you don't have to worry about building an HTTP API for the Query, handling the request on the server, or even handling and caching the responses on the client.
> Instead, simply focus on the business logic inside your Query and let Wasp take care of the rest!

Reviewer: ```suggestion
You declare queries in the `.wasp` file and implement them using NodeJS. Wasp not only runs these queries within the server's context but also creates code that enables you to call them from any part of your codebase, whether it's on the client or server side.

This means you don't have to build an HTTP API for your query, manage server-side request handling, or even deal with client-side response handling and caching.

So, just concentrate on crafting the business logic inside your query, and let Wasp handle the rest!
```

To create a Query, you must:

1. Declare the Query in Wasp using the `query` declaration.
2. Define the Query's NodeJS implementation.

After completing these two steps, you'll be able to use the Query from any point in your code.

### Declaring Queries

> We'll start by telling Wasp we wish to create a Query using a `query` declaration.
> Let's declare two Queries, one for fetching all tasks and another one for fetching tasks based on a filter (e.g., whether a task is done):

Reviewer: ```suggestion
To create a Query in Wasp, we begin with a `query` declaration. Let's declare two Queries - one to fetch all tasks, and another to fetch tasks based on a filter, such as whether a task is done:
```

  fn: import { getAllTasks } from "@server/queries.js"
}

query getFilteredTasks {
  fn: import { getFilteredTasks } from "@server/queries.js"
}
```

</TabItem>
<TabItem value="ts" label="TypeScript">

```wasp title="main.wasp"
// ...

query getAllTasks {
  fn: import { getAllTasks } from "@server/queries.js"
}

query getFilteredTasks {
  fn: import { getFilteredTasks } from "@server/queries.js"
}
```

:::warning
Even though you are using TypeScript and plan to implement this Query in `src/server/queries.ts`, you still need to import it using a `.js` extension. Wasp internally uses `esnext` module resolution, which requires importing all files with a `.js` extension. This is only needed when importing `@server` files.

Read more about ES modules in TypeScript [here](https://www.typescriptlang.org/docs/handbook/esm-node.html). If you're interested in the discussion and the reasoning behind this, read about it [in this GitHub issue](https://github.com/microsoft/TypeScript/issues/33588).
:::

</TabItem>
</Tabs>

<small>

If you want to know about all supported options for the `query` declaration, take a look at the [Options reference](#options-reference).

</small>

> Wasp Queries and their implementations don't need to (but can) have the same name. We'll keep the names identical to avoid confusion.

Reviewer: ```suggestion
The names of Wasp Queries and their implementations don't need to match, but we'll keep them the same to avoid confusion.
```

:::info
You might have noticed that we told Wasp to import Query implementations that don't yet exist. Don't worry about that for now. We'll write the implementations imported from `queries.{js,ts}` in the next section.

It's a good idea to start with the high-level concept (i.e., the Query declaration in the Wasp file) and only then deal with the implementation details (i.e., the Query's implementation in JavaScript).
:::

> After declaring a Wasp Query, two crucial things happen:

Reviewer: ```suggestion
After declaring a Wasp Query, two important things happen:
```

> - Wasp **generates a server-side NodeJS function** that shares its name with the Query. This function's interface is identical to the client-side function from the previous point.
> - Wasp **generates a client-side JavaScript function** that shares its name with the Query (e.g., `getFilteredTasks`).
>   This function takes a single optional argument - an object containing any serializable data you wish to use inside the Query.
>   Wasp will send this object over the network and pass it into the Query's implementation as its first positional argument (more on this when we look at the implementations).
>   Such an abstraction works thanks to an HTTP API route handler Wasp generates on the server, which calls the Query's NodeJS implementation under the hood.

Reviewer: ```suggestion
- Wasp **generates a server-side NodeJS function** that shares its name with the Query.
  
  This function's interface is identical to the client-side function from the previous point.

- Wasp **generates a client-side JavaScript function** that shares its name with the Query (e.g., `getFilteredTasks`).

    This function takes a single optional argument - an object containing any serializable data you wish to use inside the Query.
  Wasp will send this object over the network and pass it into the Query's implementation as its first positional argument (more on this when we look at the implementations).
  Such an abstraction works thanks to an HTTP API route handler Wasp generates on the server, which calls the Query's NodeJS implementation under the hood.
```

This section feels very heavy with details 🤯

Wasp uses [superjson](https://github.com/blitz-js/superjson) under the hood. In other words, you don't need to limit yourself to only sending and receiving JSON payloads.

Send and receive any superjson-compatible payload (e.g., Dates, Sets, Lists, circular references, etc.) and let Wasp take care of the (de)serialization.
:::

</TabItem>
<TabItem value="ts" label="TypeScript">

```ts title="src/server/queries.ts"
import { GetAllTasks, GetFilteredTasks } from "@wasp/queries/types";

type Task = {
  id: number;
  description: string;
  isDone: boolean;
};

// our "database"
const tasks: Task[] = [
  { id: 1, description: "Buy some eggs", isDone: true },
  { id: 2, description: "Make an omelette", isDone: false },
  { id: 3, description: "Eat breakfast", isDone: false },
];

// You don't need to use the arguments if you don't need them
export const getAllTasks: GetAllTasks<void, Task[]> = () => {
  return tasks;
};

// The 'args' object is something sent by the caller (most often from the client)
export const getFilteredTasks: GetFilteredTasks<
  Pick<Task, "isDone">,
  Task[]
> = (args) => {
  const { isDone } = args;
  return tasks.filter((task) => task.isDone === isDone);
};
```

> Wasp automatically generates the types `GetTasks` and `GetFilteredTasks` based on the declarations in your Wasp file:
> 
> - `GetTasks` is a generic type that Wasp automatically generated based on the Query declaration for `getTasks`.
> - `GetFilteredTasks` is a generic type that Wasp automatically generated based on the Query declaration for `getFilteredTasks`.
> 
> You can use these types to specify the Query's input and output types.
> 
> The Query `getTasks` doesn't expect any arguments (its input type is `void`), but it does return an array of tasks (its output type is `Task[]`).
> 
> The Query `getFilteredTasks`, on the other hand, expects an object of type `{ isDone: boolean }`. We've derived this type from the type `Task`.
> 
> Annotating the Queries is optional, but highly recommended. Doing so enables **full-stack type safety**. We'll see what this means when calling the Query from the client.
> 
> :::tip
> Wasp uses [superjson](https://github.com/blitz-js/superjson) under the hood. In other words, you don't need to limit yourself to only sending and receiving JSON payloads.
> 
> Send and receive any superjson-compatible payload (e.g., Dates, Sets, Lists, circular references, etc.) and let Wasp take care of the (de)serialization.
> 
> As long as you're annotating your Queries with correct automatically generated types, TypeScript ensures your payloads are valid (i.e., that Wasp knows how to serialize and deserialize them).
> :::

Reviewer: ```suggestion
Wasp automatically generates the types `GetTasks` and `GetFilteredTasks` based on your Wasp file's declarations:

- `GetTasks` is a generic type automatically generated by Wasp, based on the Query declaration for `getTasks`.
- `GetFilteredTasks` is also a generic type automatically generated by Wasp, based on the Query declaration for `getFilteredTasks`.

You can utilize these types to define the input and output types for your Query.

For instance, the Query `getTasks` doesn't expect any arguments (its input type is `void`), but it does return a list of tasks (its output type is `Task[]`).

On the other hand, the Query `getFilteredTasks` expects an object of type `{ isDone: boolean }`. This type is derived from the `Task` type.

While annotating the Queries is optional, it's highly recommended. Doing so enables **full-stack type safety**. We'll explore what this means when we discuss calling the Query from the client.

:::tip
Wasp uses [superjson](https://github.com/blitz-js/superjson) under the hood. This means you're not limited to only sending and receiving JSON payloads.

You can send and receive any superjson-compatible payload (like Dates, Sets, Lists, circular references, etc.) and let Wasp handle the (de)serialization.

As long as you're annotating your Queries with the correct automatically generated types, TypeScript ensures your payloads are valid (i.e., Wasp knows how to serialize and deserialize them).
:::
```

</TabItem>
</Tabs>

<small>

For a detailed explanation of the Query definition API (i.e., arguments and return values), check the [Options reference](#options-reference).

</small>

### Using Queries

To use a Query, you can import it from `@wasp` and call it directly. As mentioned, the usage doesn't change depending on whether you're on the server or the client:

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

> ```javascript

Reviewer: I would add `title=client/SomePage.jsx` so it's clear that it's client code

import getAllTasks from "@wasp/queries/getAllTasks.js";
import getFilteredTasks from "@wasp/queries/getFilteredTasks.js";

// ...

const allTasks = await getAllTasks();
const doneTasks = await getFilteredTasks({ isDone: true });
```

</TabItem>
<TabItem value="ts" label="TypeScript">

```typescript
import getAllTasks from "@wasp/queries/getAllTasks.js";
import getFilteredTasks from "@wasp/queries/getFilteredTasks.js";

// TypeScript automatically infers the return values and type-checks
// the payloads.
const allTasks = await getAllTasks();
const doneTasks = await getFilteredTasks({ isDone: true });
```

</TabItem>
</Tabs>

:::caution

> Wasp will not stop you from importing a Query's NodeJS implementation from `./queries.{js,ts}` and calling it directly.

Reviewer: Is this even possible? `queries.ts` is server-side code?

Reviewer: I would maybe delete this note completely, as it comments on some edge case.

However, it's best if you avoid doing so. Besides losing all useful features a Wasp Query provides (e.g., Entity injection), the raw Query will expect you to provide a context as a second argument (which is cumbersome).
:::

#### The `useQuery` hook

When using Queries on the client, you can make them reactive with the `useQuery` hook.

> This hook comes bundled with Wasp and is a thin wrapper around the `useQuery` hook from [_react-query_](https://github.com/tannerlinsley/react-query), the only difference being in not having to supply the key (Wasp does this automatically under the hood).

Reviewer: ```suggestion
This hook comes bundled with Wasp and is a thin wrapper around the `useQuery` hook from [_react-query_](https://github.com/tannerlinsley/react-query). The only difference is that you don't need to supply the key - Wasp handles this for you automatically.
```

Here's an example of calling the Queries using the `useQuery` hook:

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```jsx
import React from "react";

> import { Task } from "@wasp/entities";

Reviewer: Importing `Task` in Javascript code

import { useQuery } from "@wasp/queries";
import getAllTasks from "@wasp/queries/getAllTasks";
import getFilteredTasks from "@wasp/queries/getFilteredTasks";

const MainPage = () => {
  const { data: allTasks, error: error1 } = useQuery(getAllTasks);
  const { data: doneTasks, error: error2 } = useQuery(getFilteredTasks, {
    isDone: true,
  });

  if (error1 !== null || error2 !== null) {
    return <div>There was an error</div>;
  }

  return (
    <div>
      <h2>All Tasks</h2>
      {allTasks && allTasks.length > 0
        ? allTasks.map((task) => <Task key={task.id} {...task} />)
        : "No tasks"}

      <h2>Finished Tasks</h2>
      {doneTasks && doneTasks.length > 0
        ? doneTasks.map((task) => <Task key={task.id} {...task} />)
        : "No finished tasks"}
    </div>
  );
};

const Task = ({ description, isDone }: Task) => {
  return (
    <div>
      <p>
        <strong>Description: </strong>
        {description}
      </p>
      <p>
        <strong>Is done: </strong>
        {isDone ? "Yes" : "No"}
      </p>
    </div>
  );
};

export default MainPage;
```

</TabItem>
<TabItem value="ts" label="TypeScript">

> ```tsx

Reviewer: `title=filename`

        {isDone ? "Yes" : "No"}
      </p>
    </div>
  );
};

export default MainPage;
```

Notice how you don't need to annotate the Query's return value type. Wasp automatically infers the from the Query's backend implementation. This is **full-stack type safety**: the types on the client always match the types on the server.

</TabItem>
</Tabs>

<small>

For a detailed specification of the `useQuery` hook, check the [Options Reference](#options-reference).

</small>

### Error Handling

For security reasons, all exceptions thrown in the Query's NodeJS implementation are sent to the client as responses with the HTTP status code `500`, with all other details removed.
Hiding error details by default helps against accidentally leaking possibly sensitive information over the network.

If you do want to pass additional error information to the client, you can construct and throw an appropriate `HttpError` in your implementation:

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```js title=src/server/queries.js
import HttpError from "@wasp/core/HttpError.js";

export const getAllTasks = async (args, context) => {
  throw new HttpError(

>     403, // status coce

Reviewer: ```suggestion
    403, // status code
```

    "You can't do this!", // message
    { foo: "bar" } // data
  );
};
```

</TabItem>
<TabItem value="ts" label="TypeScript">

```ts title=src/server/queries.ts
import { GetAllTasks } from "@wasp/queries/types";
import HttpError from "@wasp/core/HttpError.js";

export const getAllTasks: GetAllTasks = async (args, context) => {
  throw new HttpError(
    403, // status coce
    "You can't do this!", // message
    { foo: "bar" } // data
  );
};
```

</TabItem>
</Tabs>

> If the status code is `4xx`, the client will receive a response object with the corresponding `.message` and `.data` fields and rethrow the error (with these fields included).
> To prevent information leakage, the server won't forward these fields for any other HTTP status codes.

Reviewer: ```suggestion
If the status code is 4xx, the client will receive a response object with the corresponding .message and .data fields, and it will rethrow the error (including these fields). To prevent the leakage of information, the server won't forward these fields for any other HTTP status codes.
```

export const getAllTasks: GetAllTasks<void, Task[]> = async (args, context) => {
  return context.entities.Task.findMany({});
};

export const getFilteredTasks: GetFilteredTasks<
  Pick<Task, "isDone">,
  Task[]
> = async (args, context) => {
  return context.entities.Task.findMany({
    where: { isDone: args.isDone },
  });
};
```

Again, annotating the Queries is optional, but greatly improves **full-stack type safety**.

</TabItem>
</Tabs>

The object `context.entities.Task` exposes `prisma.task` from [Prisma's CRUD API](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/crud).

## Options Reference

### Declaring Queries

The `query` declaration supports the following fields:

- `fn: ServerImport` <Required />

  The import statement of the Query's NodeJs implementation.

- `entities: [Entity]`

  A list of entities you wish to use inside your Query.
  For instructions on using Entities in Queries, take a look at [the guide](#using-entities-in-queries).

#### Example

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

Declaring the Query:

```wasp
query getFoo {
    fn: import { getFoo } from "@server/queries.js"
    entities: [Foo]
}
```

> Makes the following import on both the server and the client:

Reviewer: Reads weird, "Makes the following import", does it import it somewhere or what?

We mention that it is available on the server, but don't really give them an example of importing a query on the server (I don't think we should give it) so maybe no mention server here?

> ```js

Reviewer: I would wrap all code blocks with tabs even though the content is the same.

import getFoo from "@wasp/queries";
```

</TabItem>
<TabItem value="ts" label="TypeScript">

Declaring the Query:

```wasp
query getFoo {
    fn: import { getFoo } from "@server/queries.js"
    entities: [Foo]
}
```

Makes the following imports available on both the server and the client:

```ts
import getFoo from "@wasp/queries";
```

As well as the following type import on the server:

```ts
import type { GetFoo } from "@wasp/queries/types";
```

</TabItem>
</Tabs>

### Implementing Queries

The Query's implementation is a NodeJS function that takes two arguments (it can be an `async` function if you need to use the `await` keyword).
Since both arguments are positional, you can name the parameters however you want, but we'll stick with `args` and `context`:

1. `args` (type depends on the Query)

>    An object containing all the arguments (i.e., payload) **passed to the Query by the caller** (e.g., filtering conditions).

Reviewer: "the data that is passed in when calling the query"

   Check [the usage examples](#using-the-query) to see how to pass this object to the Query.

2. `context` (type depends on the Query)

>    An additional context object **passed into the Query by Wasp**. This object contains user session information, as well as information about entities. Check the [section about using entities in Queries](#using-entities-in-queries) to see how to use the `context` object.

Reviewer: Maybe link to Using auth for to see `context.user` usage

<ShowForTs>

Afer you [declare the query](#declaring-queries-in-wasp), Wasp generates a generic type you can use when defining its implementation.
For the Query declared as `getSomething`, the generated type is called `GetSomething`:

```ts
import { GetSomething } from "@wasp/queries/types";
```

It expects two (optional) type arguments:

1.  `Input`

    The type of the `args` object (i.e., the Query's input payload). The default value is `never`.

2.  `Output`

    The type of the Query's return value (i.e., the Query's output payload). The default value is `unknown`.

The defaults were chosen to make the type signature as permissive as possible. If don't want your Query to take/return anything, use `void` as a type argument.

</ShowForTs>

> #### Example

Reviewer: This example shows usage of the generic type, but it offers a Javascript impl where that doesn't make sense (there are no types used).

I would suggest wrapping the whole section, starting with "Afer you [declare the query](#declaring-queries-in-wasp), Wasp generates a generic type you can use when defining its implementation." in `<ShowForTs>` and only have the example for TS.

Author: That's exactly what I did, but forgot to import it :upside_down_face:

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

The following Query:

```wasp
query getFoo {
    fn: import { getFoo } from "@server/queries.js"
    entities: [Foo]
}
```

Expects to find a named export `getFoo` from the file `src/server/queries.js`

```js title=queries.js
export const getFoo = (args, context) => {
  // implementation
};
```

</TabItem>
<TabItem value="ts" label="TypeScript">

The following Query:

```wasp
query getFoo {
    fn: import { getFoo } from "@server/queries.js"
    entities: [Foo]
}
```

Expects to find a named export `getFoo` from the file `src/server/queries.js`

You can use the generated type `GetFoo` and specify the Query's inputs and outputs using its type arguments.

```ts title=queries.ts
import { GetFoo } from "@wasp/queries/types";

type Foo = // ...

export const getFoo: GetFoo<{ id: number }, Foo> = (args, context) => {
  // implementation
};
```

In this case, the Query expects to receive an object with an `id` field of type `number` (this is the type of `args`), and return a value of type `Foo` (this must match the type of the Query's return value).

</TabItem>
</Tabs>

> ### The `useQuery` Hook

Reviewer: Link to an example above or offer it inline
